design/authn-oidc.md

```diff
@@ -0,0 +1,121 @@
+  - [Affected Components](#affected-components)
+  - [Test Plan](#test-plan)
+    - [Test Cases](#test-cases)
+      - [Unit Tests](#unit-tests)
+      - [E2E tests](#e2e-tests)
+      - [Security testing](#security-testing)
+  - [Logs](#logs)
+  - [Documentation](#documentation)
+  - [Version update](#version-update)
+  - [Security](#security)
+  - [Audit](#audit)
+  - [Solution Review](#solution-review)
+
+## Useful Links
+
+| Link | Private |
+|------|:-------:|
+| [Conjur Docs: OIDC Authentication](https://docs.conjur.org/Latest/en/Content/OIDC/OIDC.htm) | No |
+| [Solution Design: Authn-LDAP Support in CLI](authn-ldap.md) | No |
+| [MFA for Conjur UI - Architecture](https://ca-il-confluence.il.cyber-ark.com/display/rndp/MFA+for+Conjur+UI+-+Architecture) | Yes |
+| [MFA Customer Documentation Changes](https://ca-il-confluence.il.cyber-ark.com/display/rndp/MFA+Customer+Documentation+Changes+-+Dev+Content) | Yes |
+
+## Overview
+
+Conjur supports various forms of authentication, including OIDC (OpenID Connect). OIDC is unique in that requires the user to
+navigate to a third party website (SSO provider) and authenticate there, often using MFA. The user then receives a token from
+the SSO provider and passes this to Conjur, which verifies that the token is signed correctly by the trust SSO provider.
+This document describes the design and implementation of the OIDC authentication support in the new Python-based Conjur CLI.
+
+### Motivation
+
+OIDC support was recently added to the Conjur UI (Enterprise only), initially to support SSO with Okta. We want to extend this
+support to the CLI so developers using the CLI can authenticate to Conjur in the CLI with OIDC as they will in the Conjur UI.
+
+
+## Solution
+
+### User Experience
+
+The `conjur init` command should support OIDC authentication as an authentication type. It should be possible to specify
+the authentication type `oidc` using the existing `-t` or `--authn-type` option.
+When using `oidc`, a `--service-id` option should be mandatory.
+
+*Open question*: Currently, if the `--service-id` option is specified, then `--authn-type` defaults to `ldap`. Should we make this required
+without a default value instead?
+
+| Option | Acceptable Values | Default | Description |
+| ------ | ----------------- | ------- | ----------- |
+| `--authn-type` / `-t` | `authn`, `ldap`, `oidc` | ??? | The authentication method to use when connecting to Conjur |
+| `--service-id` | `<service_id>` | N/A | The service_id of the OIDC service to use |
+
+The `conjur login` command should either:
+
+1. Open a browser to the OIDC URL configured for the SSO provider
+2. Print the URL (and/or a QR code) to be copied to a browser when running on a text-only machine
+
+The user must then authenticate to the SSO provider and receive a JWT which will be stored in the user's keychain or in the .netrc file.
+*TODO: How will the CLI get the JWT token from the browser?*
```

Comment:
> *TODO: How will the CLI get the JWT token from the browser?*

This is the million dollar question to me. Generally the OIDC provider would make a callback containing the authorization code via the `redirect_uri` that is provided in the initial authn request. How will it work if we aren't running a webserver somewhere that we can retrieve the code from?

It looks like there are some flows that don't require redirects, i.e. [the resource owner password flow](https://developer.okta.com/docs/concepts/oauth-openid/#resource-owner-password-flow) so I'm assuming we are limited to those

Reply:
Good point. One option I think I've seen apps use is to start a web server on localhost and set that as the redirect URL. That way we can get the JWT and then pass it along to the Conjur server and we can use the [Authorization Code with PKCE flow](https://developer.okta.com/docs/guides/implement-grant-type/authcodepkce/main/) which is recommended. I'll add this to the document but I'm definitely interested in other approaches.

Reply:
Looking around online, it seems that popping open a browser is the way to go. See for example [this tutorial](https://medium.com/@balaajanthan/openid-flow-from-a-cli-ac45de876ead) and [this sample code](https://github.com/puiterwijk/python-openidc-client/blob/master/openidc_client/__init__.py#L499). If we go this route, OIDC will not be supported on remote servers.